### ISSUE TITLE
Running Llama2 models locally on Windows10

### ISSUE THREAD

**Issue by BoussabatWael** [facebookresearch/llama#673]
I got approval from meta, then I downloaded all meta Llama2 models locally(I followed all steps and everything was fine). 
I tried to run the model 7B using this command “torchrun --nproc_per_node 1 example_text_completion.py --ckpt_dir llama-2-7b --tokenizer_path tokenizer.model --max_seq_len 128 --max_batch_size 4” (as mentioned on Llama2 GitHub), but when I run this command I got this error “Distributed package doesn’t have NCCL built in”, So I tried to download NCCL 2.16.5 that could support cuda 11.8, but still not working.

** My environment :
- Windows10, Nvidia Geforce RTX 3090
- CUDA11.8
- torch 2.0.1+cu118
- When I run "torch.__version __" I got 2.0.1+cu118
- When I run “torch.cuda.is_available()” I got True
- When I run “nvcc --version” I got Build cuda_11.8.r11.8

These screenshots may help to understand more the problem :

![Capture](https://github.com/facebookresearch/llama/assets/76820694/dd238e89-8439-431e-94a9-0ef7d1739cc2)
![Capture1](https://github.com/facebookresearch/llama/assets/76820694/de16b40a-ca47-4145-bc0a-4719f8444c4f)
![Capture2](https://github.com/facebookresearch/llama/assets/76820694/f7f7a260-76ef-4eee-84e3-0c548925d2f5)
![Capture5](https://github.com/facebookresearch/llama/assets/76820694/5127a15f-917c-49af-a9af-89b8fe4053f4)
![Capture6](https://github.com/facebookresearch/llama/assets/76820694/5e593dc6-cc78-43b5-85bc-5cede0d84be5)
![Capture7](https://github.com/facebookresearch/llama/assets/76820694/9973a83d-dc0a-410d-802c-3897a104e04a)
 
I want to find a way to run the script. I’m blocked for 3 days, please help if there a solution. 
Thanks in advance.

**Comment by EmanuelaBoros** [comment]
What does:

```
torch.cuda.nccl.is_available(torch.randn(1).cuda())
torch.cuda.nccl.version()
```
return?

**Comment by oneengineer** [comment]
Why running locally requires nccl?
They return false or raising error.

**Comment by BoussabatWael** [comment]
@EmanuelaBoros thank you for replying
when I run those commands I got this :
![WhatsApp Image 2023-08-17 at 10 28 44 AM](https://github.com/facebookresearch/llama/assets/76820694/db81dbb5-3b2f-44fd-a0f4-d10410829859)
![WhatsApp Image 2023-08-17 at 10 28 44 AM-2](https://github.com/facebookresearch/llama/assets/76820694/de5b4c02-550e-457b-b70b-053621382dee)
![WhatsApp Image 2023-08-17 at 10 28 44 AM-3](https://github.com/facebookresearch/llama/assets/76820694/a12959e7-648e-4a52-a1c1-2e165c757bdb)

**Comment by BoussabatWael** [comment]
@oneengineer thank you for replying

I really have no idea why nccl required.

**Comment by BoussabatWael** [comment]
@EmanuelaBoros And still run it with the same command with the same parameters  ?!

**Comment by BoussabatWael** [comment]
the solution of "robertofuentesr" here [https://github.com/facebookresearch/llama/issues/699#issuecomment-1688777017](url) works for me only with 7b and 7b-chat models, I want to know how can I run the 70b and 70b-chat.
Thank you in advance.

**Comment by EmanuelaBoros** [comment]
@BoussabatWael Great that you found the solution. However, for running 70b, one GPU is not sufficient (7b has one `.pth` file, while `70b` has 8).

**Comment by BoussabatWael** [comment]
> @BoussabatWael Great that you found the solution. However, for running 70b, one GPU is not sufficient (7b has one `.pth` file, while `70b` has 8).

So I need to implement 8 GPUs right ? and for running 13b I need 2 GPUs because the model has two consolidates. 
Thank you everyone.

**Comment by BoussabatWael** [comment]
@EmanuelaBoros Hi again,
I have Nvidia RTX 3090 and the 7b and 7b-chat models running successfully like I said before, now I want to run the 70b and 70-chat models on the same GPU, how can I do this please.
thanks in advance.

**Comment by EmanuelaBoros** [comment]
@BoussabatWael You can set nproc-per-node as the number of GPUs. For running 70b, you would need around 80GB GPU.

**Comment by BoussabatWael** [comment]
@EmanuelaBoros thank you for replying,
I have 2 GPUs of 80GB each, so when I run the model I use this command  "torchrun --nproc_per_node 2 example_text_completion.py --ckpt_dir llama-2-70b --tokenizer_path tokenizer.model --max_seq_len 128 --max_batch_size 4" ??? Here the  --nproc_per_node equal 2 and it's the number of GPUs that will run together, but the model 70b has 8 consolidate .pth, it will not raise an exception "AssertionError: Loading a checkpoint for MP=8 but world size is 2" ???? because I think the --nproc_per_node also depends of the number of .pth files of the model.


Thank you.

**Comment by EmanuelaBoros** [comment]
@BoussabatWael indeed, nproc_per_node needs to be set to the MP value - which commonly is the number of gpus. Did you try then setting nproc_per_node to 8?

**Comment by BoussabatWael** [comment]
@EmanuelaBoros  that's what I'm asking about, to run the llama2-70b I should set the nproc_per_node to 8 it will not raise "AssertionError: Loading a checkpoint for MP=8 but world size is 2" ???

**Comment by BoussabatWael** [comment]
@EmanuelaBoros Okay I will give it a try.

**Comment by EmanuelaBoros** [comment]
@BoussabatWael When you set up nproc_per_node = 8, it will create 8 processes that will instantiate llama from https://github.com/facebookresearch/llama/blob/1a240688810f8036049e8da36b073f63d2ac552c/llama/generation.py#L51 with local_rank from 0 to 7 (for each GPU). 

**Tested on my side:** to force it to load it on just 2 GPUs, run it with nproc_per_node=2 and comment out the line that checks the number of world_size https://github.com/facebookresearch/llama/blob/1a240688810f8036049e8da36b073f63d2ac552c/llama/generation.py#L79.

**Comment by haomengz** [comment]
@EmanuelaBoros Thanks for the note. I have tried this but facing the size mismatch problem. I am using 2 GPUs for 70b model. It seems that the loaded model shape will be 4 times larger than the saved model for many layers. Have you met similar problem?
